### PR TITLE
[watermarking] Dual-key gumbel-max watermarking for speculative decoding support

### PR DESCRIPTION
**Pull request by TQCB**
This PR includes work on dual-key gumbel-max watermarking, as well as generation-time context deduplication (taken from https://github.com/vllm-project/vllm/pull/56233). I've added the PR description of context deduplication to this PR as well, for the sake of future readers.

### Purpose
Text watermarking in vLLM (see https://github.com/vllm-project/vllm/pull/54053) currently is not compatible with speculative decoding. This PR proposes multiple changes to resolve this:

- `"allow_target_only_watermarking": true` allows the use of speculative decoding with watermarking, but will not watermark accepted draft tokens: this will dilute the watermarking signal in proportion to the share of output tokens that are accepted drafts
- `dual_key_gumbel` algorithm is a variation of the Gumbel-max algorithm, where we use different keys to watermark our generations: this allows for watermarking of draft tokens, and also for improved diversity of outputs when using watermarking

Dual-key watermarking allows to support speculative decoding by using algorithm 6 of [SynthID-Text's supplementary material](https://media.springernature.com/original/springer-static/esm/art%3A10.1038%2Fs41586-024-08025-4/MediaObjects/41586_2024_8025_MOESM1_ESM.pdf). To watermark the draft model's generations, without inducing a loss in acceptance rate (see algorithm 5):

```
Given distinct watermarking keys k^T and k^D, a target model p, a draft model q, a watermarked target model p_wm(.;k^T) and draft model q_wm(.;k^D)

1. We sample draft tokens from q_wm
2. We calculate target logits for drafts using p
3. We accept tokens with probability r < min(1, q(x) / p(x))
4. If rejected, we sample from residual(p, q)(.;k^T)
5. If all tokens are accepted, we sample our bonus token from p_wm
```

Note how:
- all draft tokens are watermarked by k^D
- all target tokens are watermarked by k^T (residual, and bonus)
- acceptance rate uses unwatermarked distributions, and is identical to acceptance rate without watermarking

For justification of using distinct keys to preserve non-distortion, see section K.13 of [SynthID-Text's supplementary material](https://media.springernature.com/original/springer-static/esm/art%3A10.1038%2Fs41586-024-08025-4/MediaObjects/41586_2024_8025_MOESM1_ESM.pdf).

### Test Plan
Measuring impact, using standard models and methods, on:
- inference performance (in particular the acceptance rate)
- detectability of watermark

### Test Result
Results for commit `97a264006f3d7ed9999fbd9b93319b780ac0631d` on 1x H100 per arm.

GSM8K, 300 five-shot prompts, temperature 1.0, maximum 1,024 output tokens, and Qwen3.5 MTP with three probabilistically sampled draft tokens.

#### Inference performance

| Model | Disabled output tokens | Dual-key output tokens | Disabled tok/s | Dual-key tok/s | Delta throughput |
|---|---:|---:|---:|---:|---:|
| Qwen3.5-2B | 59,769 | 72,141 | 7,366.0 | 8,254.0 | +12.1% |
| Qwen3.5-27B | 102,555 | 101,330 | 1,268.0 | 1,215.9 | -4.1% |

The generated length distributions differ, especially for the 2B model, so these are end-to-end workload measurements rather than an isolated measure of watermarking overhead.

#### Speculative decoding

| Model | Watermark | Draft acceptance rate | Mean acceptance length |
|---|---|---:|---:|
| Qwen3.5-2B | Disabled | 69.6% | 3.09 |
| Qwen3.5-2B | Dual-key Gumbel | 79.9% | 3.40 |
| Qwen3.5-27B | Disabled | 81.1% | 3.43 |
| Qwen3.5-27B | Dual-key Gumbel | 80.9% | 3.43 |

Per-request speculative metrics were only populated for requests that performed speculative steps. The table aggregates the accepted and drafted token counts from those records.

#### Quality

| Model | Watermark | Accuracy | Mean completion length | Median completion length | Repeated 4-token context rate (micro) | Reached 1,024-token cap |
|---|---|---:|---:|---:|---:|---:|
| Qwen3.5-2B | Disabled | 25.7% | 199.2 | 128.0 | 20.48% | 14/300 (4.7%) |
| Qwen3.5-2B | Dual-key Gumbel | 25.0% | 240.5 | 123.5 | 41.97% | 33/300 (11.0%) |
| Qwen3.5-27B | Disabled | 64.0% | 341.9 | 165.0 | 20.44% | 59/300 (19.7%) |
| Qwen3.5-27B | Dual-key Gumbel | 73.0% | 337.8 | 164.5 | 19.56% | 53/300 (17.7%) |

These are single stochastic runs, so accuracy differences should not be interpreted as causal.

#### Detectability

Detection uses the exact completion token IDs, both derived keys, context deduplication, and a threshold of p ≤ 0.01.

| Model | No-watermark positives | Dual-key positives | Median dual-key p-value | Mean dual-key scored tokens |
|---|---:|---:|---:|---:|
| Qwen3.5-2B | 3/300 (1.0%) | 226/300 (75.3%) | 0.000145 | 139.5 |
| Qwen3.5-27B | 3/300 (1.0%) | 69/300 (23.0%) | 0.113773 | 271.7 |

Detection improves with completion length:

| Model | <64 tokens | 64-127 | 128-255 | 256-511 | ≥512 |
|---|---:|---:|---:|---:|---:|
| Qwen3.5-2B | 40.5% | 70.9% | 80.6% | 100.0% | 100.0% |
| Qwen3.5-27B | 0.0% | 4.6% | 8.5% | 50.0% | 64.7% |

Poor detectability on the 27B generations are due to low entropy, which reduces the amount of signal our watermarking bias can embed:

| Model | Perplexity | Sampled tokens with p ≥ 0.99 |
|---|---:|---:|
| Qwen3.5-2B | 2.25 | 36.1% |
| Qwen3.5-27B | 1.19 | 72.7% |

## Purpose
Text watermarking in vLLM (see https://github.com/vllm-project/vllm/pull/54053) currently uses a short context (4 tokens, by default) before each generated token to create the pseudorandom noise that we use to watermark its sampling. It is possible for multiple positions in a sequence to have the same context, and thus have the same watermarking bias applied. This makes gumbel watermarking single-token non-distortionary instead of single-sequence non-distortionary: expectation of token distribution, over all keys, is unchanged, but expectation of sequence distributions, over all keys, is changed. In practice, this leads to:
- a reduced diversity of generations within a sequence, and
- increased likelihood of poor behaviour such a repetition and infinite generations

A solution to this is context masking, or context deduplication: if a context has already been watermarked in a sequence, we do not watermark it again. This allows for single-sequence non-distortion (see section G.3 of [SynthID-Text supplementary materials](https://media.springernature.com/original/springer-static/esm/art%3A10.1038%2Fs41586-024-08025-4/MediaObjects/41586_2024_8025_MOESM1_ESM.pdf)). An initial instinct may be that reducing the amount of tokens we watermark within a sequence will reduce the strength of our watermarking signal, but we actually already deduplicate contexts during detection to avoid biasing repetitions: we only score the first instance of a repeated context, matching what we propose for generation here. In practice however, we may imagine a case where a user will manipulate watermarked text to excluse an initial watermarking of a repeated context, only keeping the subsequent unwatermarked repetitions: this would indeed lead to a reduced watermarking signal, but this is already expected with a tampered text, and we propose that this is acceptable if it allows for increased guarantees of non-distortion.

## Test Plan
Measuring impact, using standard models and methods, on:
- degenerate generations (repetition, infinite generations)
- output length, intra-sequence diversity
- output quality
- inference performance

## Test Result
Results for commit `42908fd643fb46744f78937dc39034b02c99b172` on 1x H100

#### Microbenchmark
- ms per sample step
- mean of 100 + 20 warmup steps

| Batch | History tokens | Unmasked (ms/step) | Masked (ms/step) | Delta ms | Delta % |
|---|---|---|---|---|---|
| 1 | 2,048 | 0.036 | 0.110 | +0.073 | +201.2% |
| 1 | 8,192 | 0.036 | 0.111 | +0.075 | +208.7% |
| 1 | 32,768 | 0.037 | 0.110 | +0.073 | +197.5% |
| 32 | 2,048 | 0.041 | 0.110 | +0.069 | +166.6% |
| 32 | 8,192 | 0.041 | 0.110 | +0.069 | +165.6% |
| 32 | 32,768 | 0.042 | 0.111 | +0.069 | +166.8% |
| 256 | 2,048 | 0.260 | 0.351 | +0.092 | +35.3% |
| 256 | 8,192 | 0.260 | 0.482 | +0.222 | +85.5% |
| 256 | 32,768 | 0.260 | 0.483 | +0.223 | +86.0% |
| 1,024 | 2,048 | 1.003 | 1.280 | +0.276 | +27.6% |
| 1,024 | 8,192 | 1.004 | 1.793 | +0.790 | +78.7% |
| 1,024 | 32,768 | 1.003 | 1.792 | +0.789 | +78.6% |

#### E2E
Qwen3.5-2B

| Model | Batch | Output tokens | Unmasked tok/s | Masked tok/s | Δ throughput |
|---|---|---|---|---|---|
| Qwen3.5-2B | 1 | 2,048 | 401.4 | 408.9 | +1.88% |
| Qwen3.5-2B | 32 | 2,048 | 9,867.9 | 9,318.3 | -5.57% |
| Qwen3.5-2B | 1 | 8,192 | 402.9 | 422.7 | +4.92% |
| Qwen3.5-2B | 32 | 8,192 | 8,487.2 | 8,675.3 | +2.22% |

Qwen3.5-27B (mean of 3)

| Model | Batch | Output tokens | Unmasked tok/s | Masked tok/s | Δ throughput |
|---|---|---|---|---|---|
| Qwen3.5-27B | 1 | 512 | 48.2 | 48.2 | +0.02% |
| Qwen3.5-27B | 32 | 512 | 1,248.9 | 1,246.7 | -0.18% |
| Qwen3.5-27B | 1 | 2,048 | 48.3 | 48.3 | -0.05% |
| Qwen3.5-27B | 32 | 2,048 | 1,233.9 | 1,231.5 | -0.19% |
| Qwen3.5-27B | 1 | 8,192 | 48.2 | 48.2 | -0.05% |

### Quality & Diversity
Qwen3.5-2B T=1.0

| Metric | No watermark | Watermark, unmasked | Watermark, masked |
|---|---|---|---|
| Generations | 300 | 300 | 300 |
| Generations reaching the 1,024-token cap | 19 (6.3%) | 107 (35.7%) | 12 (4.0%) |
| Repeated 4-token context rate (micro) | 25.37% | 74.94% | 17.99% |
| Mean completion length (tokens) | 211.4 | 444.8 | 188.9 |
| Answer accuracy | 29.7% | 25.7% | 29.0% |

image

image

### Detectability
Detectability on the GSM8K responses

| Metric | No watermark | Watermark, unmasked | Watermark, masked |
|---|---|---|---|
| Positive at p ≤ 0.01 | 8/300 (2.7%) | 262/300 (87.3%) | 260/300 (86.7%) |
| Mean p-value | 0.51183 | 0.02107 | 0.01689 |
| Mean scored tokens | 157.8 | 111.5 | 154.9 |

### CODE REVIEW

**Comment by tomasruizt** [comment]
Hi @TQCB! Thanks for the integration with SD :rocket: Currently, it is not super easy to understand the relationship of different watermarking classes and their specific responsibilities. I would lean towards flattening the hierarchy and reducing indirection if possible. Here are some ideas meant as recommendations rather than hard requests.

| Current abstraction | Possible simplification |
|---|---|
| `WatermarkedRejectionSampler` subclass | Pass optional watermark inputs directly to `RejectionSampler`, since the subclass mainly supplies arguments to the kernel. |
| `DraftWatermarker` wrapping another watermarker | Perhaps we can extract a state object for `contexts` and `enabled` and have clients call directly the currently wrapped watermarker on the state object |
| `DualKeyGumbelWatermarker` inheriting from `GumbelWatermarker` | Prefer composition with two explicit keyed samplers, so the object doesn't simultaneously have to represent key A and the two-key algorithm. |

In general I suggest simplifying the plumbing between the watermarking components. Removing the rejection-sampler subclass seems like a relatively small starting point. The `SupportsSpeculativeDecoding` machinery seems a bit overengineered for the current use case, and could perhaps be simplified.

Let me know what you think. Happy to look at performance metrics tomorrow!

**Comment by TQCB** [comment]
@tomasruizt Thanks for the comment! Haven't looked super far into this but I think:
- we can probably remove `WatermarkedRejectionSampler`, especially with the fused approach we have
- also no objection for `DualKeyGumbelWatermarker` composition approach
- I think `DraftWatermarker` handles quite a lot that we can't easily move to a state object, I think this will be difficult to flatten without a tonne of duplication
- as for the protocol, I quite like the boundary it offers and would like to keep something similar but I think it can be reworked

**Comment by tomasruizt** [comment]
@TQCB In terms of high level metrics the PR is in a great state! I tested on H100.

* Correctness: I ran a full GSM8K correctness check on Qwen3.5-4B with watermarking + MTP, and the accuracy is unaffected.
* The simple e2e tests validate the detection in the combined MTP + watermarking case.
* Speed: I ran `vllm bench sweep serve` on Qwen3.5-9B + MTP with watermarking and without watermarking, concurrency 8, and the results are below. Acceptance length is unaffected, TPOT degrades a bit, and (surprisingly) TTFT improves a bit, but it might not be statistically significant. p-values are from a Welch t-test. I passed `--ignore-eos`, so cannot comment on output len.

| Check | Watermark | No-Watermark | % change vs No-Watermark | p-value |
|---|---|---|---:|---:|
| Acceptance length (tokens/step) | 3.4165 | 3.4177 | -0.033% | 0.82 |
| Time per output token (TPOT) | 3.327 ms | 3.271 ms | +1.72% | 4.9e-8 |
| Time to first token (TTFT) | 49.95 ms | 51.14 ms | -2.33% | 0.030 |


Here is the commit with the benchmarking code: https://github.com/tomasruizt/vllm/commit/5c8718118#diff-cfe2f521bfba5b72a05e55320d41bf569180041e89783353bfd56417749ae9dd

Here are the results of `vllm bench sweep serve`:  [summary.csv](https://github.com/user-attachments/files/32096933/summary.csv)

@simon-veitner-redhat FYI

**Comment by tomasruizt** [comment]
I created torch profiles for the Qwen3.5-9B + MTP with/without watermarking. They show the watermaking path takes extra time, but in line with the measurements above. No CPU-GPU syncs found.

[no-wm-rank0.1789117197867402112.pt.trace.json.gz](https://github.com/user-attachments/files/32099319/no-wm-rank0.1789117197867402112.pt.trace.json.gz)
[use-wm-rank0.1789117274024498288.pt.trace.json.gz](https://github.com/user-attachments/files/32099320/use-wm-rank0.1789117274024498288.pt.trace.json.gz)

**Comment by tomasruizt** [comment]
@TQCB This a patch to remove the `WatermarkedRejectionSampler` class https://github.com/tomasruizt/vllm/commit/99f8f31e97efe5b2c6376447952d0ce17c5c9039 It declutters a bit the `model_runner.py` file. Its makes clearer that the `RejectionSampler` is only slightly modified by watermarking, rather than being completely replaced 👍

**Comment by tomasruizt** [comment]
This is a patch to use composition in `DualKeyGumbelWatermarker`. It simplifies mostly the `create_X_watermarker()` methods and the interface with plain attributes. Deferred creation is not strictly necessary atm, right? https://github.com/tomasruizt/vllm/commit/430807bba49833e4c27a053ce6f8b6cb6d84e10c

**Comment by simon-veitner-redhat** [comment]
i am pretty sure that the low detection rate is an artifact of 27B model remembering the answers to GSM8K.
see below for experiment i ran with claude

---

I looked into the 27B detection numbers. TL;DR: they reproduce, and the cause is how little entropy the 27B leaves in five-shot GSM8K answers, not the dual-key scheme, MTP, or the base feature. On ordinary prose the 27B is detected in 99–100% of completions. Not a blocker; the description and docs should say this instead of guessing at it.

**Reproduction.** Same setup as the description on the PR head (`e7a3524d3`), 150 prompts, this time retaining logprobs at every position.

| | 2B | 27B |
|---|---|---|
| detected (this run / description) | 72.7% / 75.3% | 22.7% / 23.0% |
| per-token evidence over the null | 0.67 | 0.12 |
| sampled tokens with p(x) > 0.999 | 18% | 67% |

**Why.** For Gumbel-max the expected detector score of a token depends only on the probability the sampler gave it, `f(p) = Σ_k 1/(k(1+kp))`, which is the null value at p = 1. The observed per-token scores match this prediction for both models within 2–4%, and within one standard error in every probability bin. Two thirds of the 27B's GSM8K tokens are near-certain and carry no evidence, so the dual detector needs ~1000 scored tokens for 50% power on that task, against ~50 for the 2B.

**It is the text, not the model.** Same engines, same key, same MTP setting, 150 prompts each, controls at 0–2 of 150:

| prompt set | 2B | 27B | 27B evidence/token |
|---|---|---|---|
| C4 news continuation | 150/150 | 149/150 | 1.88 |
| WritingPrompts stories | 150/150 | 149/150 | 0.72 |
| custom stories (unmemorizable, generated) | 50/50 | 50/50 | 0.67 |
| custom explanations | 50/50 | 50/50 | 0.58 |
| custom fixed-answer lists (times tables, counting) | 1/50 | 0/50 | ≈ 0 |
| five-shot GSM8K | 72.7% | 22.7% | 0.13 |

The fixed-answer tier is GSM8K in pure form: both models write it at p(x) > 0.99 on 87–99% of positions and neither is detected, at any size. Two more controls: the 2B at temperature 0.6 drops to 17%, the 27B's rate; the 27B without MTP is at 15%, so spec decode is a 10–20% modifier, not the cause.

**Comment by simon-veitner-redhat** [comment]
please cherry pick 
https://github.com/simon-veitner-redhat/vllm/commit/c397812d4788f7ed3daabfe4934c895bb9d60253
for fix of CI

**Review by LucasWilkinson — APPROVED** [review]
LGTM, thanks!
